### PR TITLE
A couple commits culminating in support for function pointer type inference

### PR DESCRIPTION
**Pull request by Ptival**
Still not perfect, but it's getting so large that I'd rather set this as a checkpoint.

### CODE REVIEW

File: src/Reopt/TypeInference/Header.hs [hunk 1]
```diff
@@ -145,15 +145,34 @@ parseStructUnion (C.CStruct tag _mi _mdecl _attrs n) =
...
+  typeSpecs <- mapM parseTypeDecls decls
+  args <- mapM (parseType emptyQualMods) typeSpecs
+  parseTypeDerivedDecl rest $! FunPtrAnnType tp args
+  -- errorAt n $ "decls: " ++ show decls ++ "\nattrs: " ++ show attrs ++ "\ntp: " ++ show tp
+parseTypeDerivedDecl (C.CFunDeclr (Right (_, True)) _boop n : _) _tp = do
+  errorAt n "True"
+parseTypeDerivedDecl (C.CFunDeclr (Left {}) _boop n : _) _tp = do
+  errorAt n "Left"
+-- parseTypeDerivedDecl (C.CFunDeclr _ _ n : _) _tp = do
+--   errorAt n "Function declarations are not supported in this context."
```
Comment (RyanGlScott) [inline src/Reopt/TypeInference/Header.hs:175]: The commented-out code seems to give a more human-friendly error message... should that be used instead?

File: src/Reopt/TypeInference/Header.hs [hunk 2]
```diff
@@ -145,15 +145,34 @@ parseStructUnion (C.CStruct tag _mi _mdecl _attrs n) =
...
--- | Parser derived declarators.
+parseTypeDecl :: C.NodeInfo -> C.CDeclarationSpecifier C.NodeInfo -> CParser C.CTypeSpec
+parseTypeDecl _ (C.CTypeSpec spec) = return spec
+parseTypeDecl n _ = errorAt n "Expected type specification"
+
+parseTypeDecls :: C.CDeclaration C.NodeInfo -> CParser C.CTypeSpec
+parseTypeDecls (C.CDecl [decl] _ n) = parseTypeDecl n decl
+-- NOTE (val) So far I'm only seeing singleton lists here, not sure why.
+parseTypeDecls (C.CDecl _decls _ n) = errorAt n "Expected single type spec, investigate."
+parseTypeDecls (C.CStaticAssert _ _ n) = errorAt n "TODO"
```
Comment (RyanGlScott) [inline src/Reopt/TypeInference/Header.hs:156]: Are you intended to fill in this TODO?

File: src/Reopt/CFG/FnRep.hs [hunk 3]
```diff
@@ -234,11 +234,18 @@ data FnValue (arch :: Type) (tp :: M.Type) where
...
-  -- is the type of the argument.
+  -- | Value is a function argument.  The int should be in the range
+  -- @[0..argCount)@, and the type repr is the type of the argument.
   FnArg :: !Int -> !(TypeRepr tp) -> FnValue arch tp
+  -- | Value is a constant pointer into the executable segment
+  FnCodePointer ::
+    Macaw.MemAddr (ArchAddrWidth arch) ->
+    FnValue arch (BVType (ArchAddrWidth arch))
+  -- | Value is a constant pointer into the executable segment
+  FnTypedCodePointer ::
```
Comment (RyanGlScott) [inline src/Reopt/CFG/FnRep.hs:245]: These two constructors have the same Haddocks... say something about what makes them different?

**Review by RyanGlScott — APPROVED** [review]
I have not looked at every line of code (let me know if something deserves further scrutiny), but I like where this is going.

File: src/Reopt/TypeInference/Header.hs (hunk 1 above)
Comment (Ptival) [reply src/Reopt/TypeInference/Header.hs:175]: Oh yeah this was just for debugging purposes, removing!

**Pull request merged**